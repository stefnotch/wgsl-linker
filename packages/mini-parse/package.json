{
  "name": "mini-parse",
  "version": "0.5.0-d1",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "pr": "printenv",
    "prepublishOnly": "run-s build",
    "build": "run-s build:main build:util build:minified",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:minified": "vite build --minify --emptyOutDir=false",
    "test": "vitest",
    "organize": "organize-imports-cli tsconfig.json",
    "format": "prettier-eslint \"src/**/*.ts\" --write",
    "lint": "eslint src"
  },
  "type": "module",
  "repository": "github:wgsl-tooling-wg/wgsl-linker",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./minified": {
      "import": "./dist/minified.js",
      "types": "./dist/index.d.ts"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^22.5.0",
    "@typescript-eslint/eslint-plugin": "^8.12.2",
    "@typescript-eslint/parser": "^8.12.2",
    "@vitest/ui": "^2.1.4",
    "berry-pretty": "^0.0.5",
    "eslint": "^9.13.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "typescript": "^5.6.3",
    "vite": "^5.4.10",
    "vite-plugin-dts": "^4.3.0",
    "vite-tsconfig-paths": "^5.0.1",
    "vitest": "^2.1.4"
  }
}