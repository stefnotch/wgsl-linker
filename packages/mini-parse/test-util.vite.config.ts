/// <reference types="vitest" />
import assert from "node:assert";
import { resolve } from "path";
import { LibraryOptions, defineConfig } from "vite";
import baseViteConfig from "./vite.config.js";

const config = baseViteConfig;
assert(config.build);
config.build.emptyOutDir = false;
const lib = config.build.lib as LibraryOptions;
lib.name = "mini-parse-test-util";
lib.formats = ["es"];
lib.entry = [resolve(__dirname, "./src/test-util/index.ts")];
lib.fileName = "testUtil";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);

// {
//   plugins: [
//     tsconfigPaths(),
//     dts(), // generate
//     // visualizer({ brotliSize: true, gzipSize: true }), // generate stats.html size report
//   ],
//   build: {
//     emptyOutDir: false,
//     lib: {
//       name: "mini-parse-testing",
//       formats: ["es"],
//       entry: [resolve(__dirname, "./src/test-util/index.ts")],
//       fileName: "testUtil"
//     },
//     rollupOptions: {
//       // make sure to externalize deps that shouldn't be bundled
//       // into your library
//       external: ['vitest', 'mini-parse'],
//     },
//     minify: false,
//     sourcemap: true,
//   },
// });
