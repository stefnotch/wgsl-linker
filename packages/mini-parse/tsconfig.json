{
  "extends": ["../../tsconfig.options.json"],
  "compilerOptions": {
    "lib": ["DOM", "DOM.Iterable", "ESNext", "ESNext.AsyncIterable"],
    "paths": {
      "mini-parse": ["./src"],
      "mini-parse/test-util": ["./src/test-util"]
    }
  },
  "include": ["src/**/*", "*.config.ts"],
  "exclude": ["node_modules"]
}
