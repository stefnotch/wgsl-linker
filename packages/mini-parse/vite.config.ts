/// <reference types="vitest" />
import { resolve } from "path";
import { defineConfig } from "vite";
import dts from "vite-plugin-dts";

export default defineConfig({
  plugins: [dts()],
  test: { setupFiles: "./src/test/TestSetup.ts" },
  build: {
    lib: {
      name: "mini-parse",
      entry: [resolve(__dirname, "src/index.ts")],
    },
    minify: false,
    sourcemap: true,
  },
});
