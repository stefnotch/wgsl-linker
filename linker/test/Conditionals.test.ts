import { srcLog, _withBaseLogger } from "@wesl/mini-parse";
import { expectNoLogErr, logCatch } from "@wesl/mini-parse/test-util";
import { expect, test } from "vitest";
import { assertSnapshot } from "@std/testing/snapshot";
import { processConditionals } from "../Conditionals.ts";

test("parse #if #endif", async (ctx) => {
  const src = `
    #if foo
    fn f() { }
    #endif
    `;

  const sourceMap = processConditionals(src, { foo: true });
  expect(sourceMap.dest).toContain("fn f() { }");
  await assertSnapshot(ctx, sourceMap.entries);
});

test("parse // #if !foo", () => {
  const src = `
    // #if !foo
      fn f() { }
    // #endif 
    `;
  const { dest } = processConditionals(src, { foo: false });
  expect(dest).toContain("fn f() { }");
});

test("parse #if !foo (true)", () => {
  const src = `
    // #if !foo
      fn f() { }
    // #endif 
    `;
  expectNoLogErr(() => {
    const { dest } = processConditionals(src, { foo: true });
    expect(dest).not.toContain("fn");
    expect(dest).not.toContain("//");
  });
});

test("parse #if !foo #else #endif", () => {
  const src = `
    // #if !foo
      fn f() { notfoo(); }
    // #else
      fn g() { foo(); }
    // #endif 
    `;
  const { dest } = processConditionals(src, { foo: true });
  expect(dest).toContain("fn g()");
  expect(dest).not.toContain("fn f()");
});

test("parse nested #if", () => {
  const src = `
    #if foo

    #if bar
      fn f() { }
    #endif

    #if zap
      fn zap() { }
    #endif

      fn g() { }
    #endif 
    `;
  const { dest } = processConditionals(src, { foo: true, zap: true });
  expect(dest).toContain("fn zap()");
  expect(dest).toContain("fn g()");
  expect(dest).not.toContain("fn f()");
});

test("parse #if #endif with extra space", () => {
  const src = `
    #if foo 
    fn f() { }
    #endif
    `;

  const { dest } = processConditionals(src, {});
  expect(dest).not.toContain("fn f() { }");
});

test("parse last line", () => {
  const src = `
    #x
    y`;
  const { dest } = processConditionals(src, {});
  expect(dest).toBe(src);
});

test("srcLog with srcMap", async (ctx) => {
  const src = `
  #if !foo
  1234
  #endif`;
  const sourceMap = processConditionals(src, {});

  const { log, logged } = logCatch();
  _withBaseLogger(log, () => {
    srcLog(sourceMap, [3, 6], "found:");
  });

  await assertSnapshot(ctx, logged());
});

test("unterminated #if", async (ctx) => {
  const src = `
  #if foo
    // bar
  `;
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => {
    processConditionals(src, {});
  });
  await assertSnapshot(ctx, logged());
});

test("unterminated #else", async (ctx) => {
  const src = `
  #if foo
  #else
    // bar
  `;
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => {
    processConditionals(src, {});
  });
  await assertSnapshot(ctx, logged());
});
