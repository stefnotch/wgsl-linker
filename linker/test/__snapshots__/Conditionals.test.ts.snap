export const snapshot = {};

snapshot[`parse #if #endif 1`] = `
[
  {
    destEnd: 1,
    destStart: 0,
    src: "
    #if foo
    fn f() { }
    #endif
    ",
    srcEnd: 1,
    srcStart: 0,
  },
  {
    destEnd: 16,
    destStart: 1,
    src: "
    #if foo
    fn f() { }
    #endif
    ",
    srcEnd: 28,
    srcStart: 13,
  },
  {
    destEnd: 20,
    destStart: 16,
    src: "
    #if foo
    fn f() { }
    #endif
    ",
    srcEnd: 43,
    srcStart: 39,
  },
]
`;

snapshot[`srcLog with srcMap 1`] = `
"found:
  1234   Ln 3
  ^  ^"
`;

snapshot[`unterminated #if 1`] = `
"unmatched #if/#else
  #if foo   Ln 2
  ^"
`;

snapshot[`unterminated #else 1`] = `
"unmatched #if/#else
  #else   Ln 3
  ^"
`;
